src/components/editor/assembly.grammar

```diff
@@ -1,50 +1,182 @@
+  statement*
+}
 
-@skip { space | LineComment }
+@skip { space | comment }
 
-expression {
-  Area
+statement {
+  AreaDeclaration
 }
 
-Area { "AREA" "," Label "," "DATA" "," ( "READONLY" | "READWRITE" ) Instruction+ }
+AreaDeclaration {
+  kw<"AREA"> commaSep<areaProps> instruction*
+}
 
-Instruction {
-  InstructionGroup01 |
-  InstructionGroup02 |
-  InstructionGroup03 |
-  InstructionGroup04 |
-  InstructionGroup05 |
-  InstructionGroup06 |
-  InstructionGroup07 |
-  InstructionGroup08 |
-  InstructionGroup09
+areaProps {
+  AreaName |
+  ReadOrWriteArea |
+  CodeOrDataArea
 }
 
-InstructionGroup01 { Label? InstructionName LowReg "," HighReg }
-InstructionGroup02 { Label? InstructionName LowReg "," LowReg }
-InstructionGroup03 { Label? InstructionName LowReg "," ImmEight }
-InstructionGroup04 { Label? InstructionName LowReg "," LowReg "," LowReg }
-InstructionGroup05 { Label? InstructionName LowReg "," LowReg "," ImmThree }
-InstructionGroup06 { Label? InstructionName LowReg "," LowReg "," ImmFive }
-InstructionGroup07 { Label? InstructionName LowReg "," "[" LowReg "," LowReg "]" }
-InstructionGroup08 { Label? InstructionName LowReg "," "[" LowReg "," ImmFive "]" }
-InstructionGroup09 { Label? InstructionName LowReg "," "[" "PC" "," ImmEight "]" }
+AreaName {
+  identifier
+}
 
-@tokens {
-  ImmEight { "#" $[0-9] | "#" $[1-9] $[0-9] | "#1" $[0-9] $[0-9] | "#2" $[0-4] $[0-9] | "#25" $[0-6] }
-  ImmFive { "#" $[0-9] | "#" $[1-2] $[0-9] | "#3" $[0-2] }
-  ImmThree { "#" $[0-8] }
+ReadOrWriteArea {
+  kw<"READWRITE"> | kw<"READONLY">
+}
+
+CodeOrDataArea {
+  kw<"CODE"> | kw<"DATA">
+}
+
+instruction {
+  MachineInstruction |
+  SymbolicDeclaration |
+  PushPopInstruction |
+  BranchInstruction
+}
+
+SymbolicDeclaration {
+  VariableName kw<"EQU"> IntegerLiteral
+}
+
+MachineInstruction {
+  Label? MachineInstructionName commaSep<machineInstructionArg>
+}
+
+PushPopInstruction {
+  (kw<"PUSH"> | kw<"POP">) "{" commaSep<RegisterLiteral> "}"
+}
+
+BranchInstruction {
+  BranchInstructionWithLabel |
+  BranchInstructionWithoutLabel
+}
+
+BranchInstructionWithLabel {
+  ( 
+    kw<"B"> |
+    kw<"BL"> |
+    kw<"BEQ"> |
+    kw<"BNE"> |
+    kw<"BCS"> |
+    kw<"BHS"> |
+    kw<"BCC"> |
+    kw<"BLO"> |
+    kw<"BMI"> |
+    kw<"BPL"> |
+    kw<"BVS"> |
+    kw<"BVC"> |
+    kw<"BHI"> |
+    kw<"BLS"> |
+    kw<"BGE"> |
+    kw<"BLT"> |
+    kw<"BGT"> |
+    kw<"BLE"> |
+    kw<"BAL">
+  )
+  Label
+}
 
-  LowReg { "R" $[0-7] }
-  HighReg { "R" $[0-9] | "R1" $[0-2] | "SP" | "LR" | "PC" }
+BranchInstructionWithoutLabel {
+  ( 
+    kw<"BX"> |
+    kw<"BLX">
+  )
+  RegisterLiteral
+}
 
-  InstructionName { $[A-Z]+ }
-  Label { std.asciiLetter+ }
+MachineInstructionName {
+  kw<"MOV"> | 
+  kw<"MOVS"> |
+  kw<"ADDS"> |
+  kw<"ADD"> |
+  kw<"ADCS"> |
+  kw<"ADR"> |
+  kw<"SUBS"> |
+  kw<"SBCS"> |
+  kw<"RSBS"> |
+  kw<"MULS"> |
+  kw<"CMP"> |
+  kw<"CMN"> |
+  kw<"CMP"> |
+  kw<"ANDS"> |
+  kw<"EORS"> |
+  kw<"ORRS"> |
+  kw<"BICS"> |
+  kw<"MVNS"> |
```

Comment:
Please add kw<"MRS"> as instruction since if I implement the instruction I would not add it here

Reply:
done